### ISSUE TITLE
Thunderbolt hotplug

### ISSUE THREAD

**Issue by ex-git** [zearp/Nucintosh#26]
Tested with a thunderbolt certified hard drive. Hard drive did show up when plug in. Eject the hard drive and unplug will cause system reboot in couple secs. Seems like a common issue with Hackintosh. Maybe ThunderboltReset TbtForcePower.efi can help. If you have time and a tb device, maybe you can take a look into resolving this. Thanks.

**Comment by zearp** [comment]
I don’t have any tb3 devices, only tb2 which require a conversion cable to work with tb3. I’ve only had the chance to test some docks and they worked as expected. The NUC hotplug patch has been applied and another user reported hotplug worked for them if I recall correctly.

But of course it can depend on the products too. Some eGPU cases may not work while others do. Same applies to other tb stuff  Unfortunately tb devices cost a lot of money and I’m not in a position to just buy all kinds of tb3 devices to test with. All I can contribute is to say double check the BIOS, legacy mode has to be set. Without it weird issue happened here.

Since you have a device and run into an issue you’re the perfect candidate to test your suggestions. There’s some good documentation on patching tb on the NUC8 HC model that could be helpful. I’ll gladly add any fixes and improvements you find and test. The HC model does have a different tb chipset but it doesn’t make the three part write up any less interesting or helpful. 

https://osy.gitbook.io/hac-mini-guide/details/thunderbolt-3-fix

**Comment by zearp** [comment]
I don’t have any tb3 devices, only tb2 which require a conversion cable to work with tb3. I’ve only had the chance to test some docks and they worked as expected. The NUC hotplug patch has been applied and another user reported hotplug worked for them if I recall correctly.

But of course it can depend on the products too. Some eGPU cases may not work while others do. Same applies to other tb stuff.   Unfortunately tb devices cost a lot of money and I’m not in a position to just buy tb3 devices to test with. All I can contribute is to say double check the BIOS, legacy mode has to be set. Without it weird issues happened here.

Since you have a device and run into an issue you’re the perfect candidate to test your suggestions. There’s some good documentation on patching tb on the NUC8 HC model that could be helpful. I’ll gladly add any fixes and improvements you find and test. The HC model does have a different tb chipset but it doesn’t make the three part write up any less interesting or helpful. 

https://osy.gitbook.io/hac-mini-guide/details/thunderbolt-3-fix

**Comment by ex-git** [comment]
BIOS, legacy mode was set. I test various methods included the one use tbpatch to patch firmware. Patched firmware work but it also have some compatibility issues, and will lost hot plug if using other system. I don't use eGPU so I can't tell if eGPU hot plug work and I don't have fix or improvement in this case.

**Comment by zearp** [comment]
Hot plug is a tricky thing to get working, specially when it comes to tb. I have applied the usb-c/tb ACPI patch that fixes a lot but I guess it doesn't fix everything. I've seen some other tb related ACPI patches for our NUC models but since I couldn't test them properly I didn't end up using any of those. I was only able to test the one in use now.

Currently the EFI uses [this](https://github.com/Rashed97/Intel-NUC-DSDT-Patch/commit/47476815b52f8e4c97e8f85df158c9ab1b6ecedd) patch/rename and nothing else. On the HC NUC they sort of fixed their issues by flashing  slightly modified Mac tb firmware to the NUCs controller but the controller in the HC model is different from ours. But the principles of the patching procedure still apply. The thing is that I have no tb3 devices and only have tb2 docks that are pretty much glorified usb hubs with more bandwidth. But they do work with the overpriced Apple tb2 <-> tb3 conversion cable.

Without the patch I also wasn't able to map the tb3 usb stuff found [here](https://github.com/zearp/Nucintosh/blob/master/EFI/OC/Kexts/USBPorts.kext/Contents/Info.plist#L25-L75). The controller itself wouldn't even show up with out it. My guess is that some extra ACPI patching is needed to sort out hot plug, it may even need several depending on the device connected.

From what I gathered on forums/Github when I was setting my EFI up is that hot plugging is precarious and many times doesn't work even with ACPI patched as good as possible. Certain tb devices don't even work unless plugged in before booting up. The 2018 MacBook Pro and Mac mini have the exact same tb controller as our NUC though. So it would be interesting to get the tb firmware form one of those machine to compare. But we'll need to find someone who owns those machines and is willing to dump the firmware using a chip reader.

The patch currently applied does address some of the same things as the HC write up mentions, many things are very similar in their ACPI but not everything is the same. Someone with ACPI knowledge might be able to sort this out using the information in the write up and the patch we use + DSDT from the NUC. My ACPI skills are limited at best and I lack the required hardware to do any testing. Plus we're missing Apple's firmware for comparing. Checking how Linux does it is an option. Does hot plugging that disk work in Linux?

All in all I think in the short term you'll have to live with not being able to hot plug every tb device, some may work some may not. In the long run I'm sure this can be fixed given someone with knowledge and time wants to have a go at sorting out the ACPI patches. I'm not too confident that kexts or EFI drivers would be able to sort it out. We have to somehow 'translate' how Intel does things to how Apple wants them to be, which is best done in ACPI. I'll see if I can find those other NUC8 patches I found ages ago, can't hurt to try them out and see if it helps. But I really have no clue where I found them and it's been a while since I did the ACPI stuff for the NUC.

I checked Amazon and it looks like many enclosures on there state Thunderbolt compatible but are really just usb-c (10gbit max). Guess technically they're Thunderbolt compatible but misleading too. The cheapest real (40gbit) Intel certified tb enclosure I could find was nearly $100. My main issue is that I don't have any tb3 devices to test with and lack the expandable money to just buy some for testing purposes. Even if I would buy some enclosures chances are that some will work and some might not and the ones that don't work may require different methods to fix them. But that's just guessing based on what I've read a few months ago when I doing the ACPI patches.

tl;dr There's really not much I can do.

**Comment by zearp** [comment]
Here's someone who claims to have hot plug fully working, but it does involve flashing your tb controller firmware. Do not do this unless you have a programmer to recover from a bad flash, or if you wanna live on the edge lol.

https://www.tonymacx86.com/threads/success-intel-nuc8i5beh-opencore-6-1-working-thunderbolt-hot-plug.304830/

So, it may be needed to flash a custom firmware after all but I'm pretty sure it's also possible to get everything sorted with ACPI patches. The latter may take longer and be more complicated than simply flashing the firmware to behave more like Apple's firmware, but it is less risky.

I still haven't found those older ACPI patches for the NUC8 I found, they are probably way too old by now anyways. But I stumble into them I'll put them here. I have them saved somewhere...

**Comment by zearp** [comment]
Here's someone who claims to have hot plug fully working, but it does involve flashing your tb controller firmware. Do not do this unless you have a programmer to recover from a bad flash, or if you wanna live on the edge lol.

https://www.tonymacx86.com/threads/success-intel-nuc8i5beh-opencore-6-1-working-thunderbolt-hot-plug.304830/

So, it may be needed to flash a custom firmware after all but I'm pretty sure it's also possible to get everything sorted with ACPI patches. The latter may take longer and be more complicated than simply flashing the firmware to behave more like Apple's firmware, but it is less risky.

I still haven't found those older ACPI patches for the NUC8, they are probably way too old by now anyways. But if I stumble into them I'll put them here. I have them saved somewhere...

**Comment by ex-git** [comment]
That is the one I tried and I flashed back because it break tb3 hot plug in other os. I will just leave it as it is right now or boot with tb3 hard drive plugin so hot plug will work in that way. Thank you for looking. The cheapest certified tb3 case with 480gb drive is $87.99. In case you are interested. Search Nekteck 480GB Thunderbolt 3 in Amazon.

**Comment by zearp** [comment]
Did hot plugging and sleep work in macOS when using the modified firmware?

If sleep didn't work I could try to sort that out and hot plugging not working in Windows might be due to the fact the default Windows driver expect a different firmware and might be solved by installing Apple's bootcamp tb driver. Apple's driver should be able to talk to the modified firmware but thats something I can't test myself.

**Comment by zearp** [comment]
Did hot plugging and sleep work in macOS when using the modified firmware?

If sleep didn't work I could try to sort that out and hot plugging not working in Windows might be due to the fact the default Windows driver expects a different firmware and might be solved by installing Apple's bootcamp tb driver. Apple's driver should be able to talk to the modified firmware but thats something I can't test myself.

**Comment by ex-git** [comment]
With modified firmware hot plugging work in mac os but broke in Ubuntu. I didn't test further. I prefer ACPI patch. Maybe in the future it can be done.
